File: 03_arrays_part3/4_majority_element_2.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        e1 = e2 = None
        c1 = c2 = 0
        for num in nums:
            if c1 == 0 and e2 != num:
                e1 = num
                c1 = 1
            elif c2 == 0 and e1 != num:
                e2 = num
                c2 = 1
            elif num == e1:
                c1 += 1
            elif num == e2:
                c2 += 1
            else:
                c1 -= 1
                c2 -= 1
        c1 = c2 = 0
        for num in nums:
            if num == e1:
                c1 += 1
            elif num == e2:
                c2 += 1
        n = len(nums)
        res = []
        if e1 is not None and c1 > n // 3:
            res.append(e1)
        if e2 is not None and c2 > n // 3:
            res.append(e2)
        return res

File: 03_arrays_part3/4_majority_element_2.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> Vec<i32> {
        let (mut e1, mut e2, mut c1, mut c2) = (0, 1, 0, 0);

        for &num in &nums {
            if num == e1 {
                c1 += 1
            } else if num == e2 {
                c2 += 1
            } else if c1 == 0 {
                e1 = num;
                c1 = 1;
            } else if c2 == 0 {
                e2 = num;
                c2 = 1;
            } else {
                c1 -= 1;
                c2 -= 1;
            }
        }

        let n = nums.len();
        let (mut c1, mut c2) = (0, 0);
        for &num in &nums {
            if num == e1 {
                c1 += 1;
            }
            if num == e2 {
                c2 += 1;
            }
        }
        let mut res: Vec<i32> = vec![];
        if c1 > n / 3 {
            res.push(e1)
        }
        if c2 > n / 3 {
            res.push(e2)
        }

        res
    }
}
